src/pykeen/nn/utils.py

```diff
@@ -90,3 +90,75 @@ def encode_all(
             [self(batch) for batch in chunked(tqdm(labels), batch_size)],
             dim=0,
         )
+
+
+def iter_matrix_power(matrix: torch.Tensor, max_iter: int) -> Iterable[torch.Tensor]:
+    """
+    Iterate over matrix powers.
+
+    :param matrix: shape: `(n, n)`
+        the square matrix
+    :param max_iter:
+        the maximum number of iterations.
+
+    :yields: increasing matrix powers
+    """
+    a = matrix
+    for _ in range(max_iter):
+        yield a
+        # if the sparsity becomes too low, convert to a dense matrix
+        # note: this heuristic is based on the memory consumption,
+        # for a sparse matrix, we store 3 values per nnz (row index, column index, value)
+        # performance-wise, it likely makes sense to switch even earlier
+        # `torch.sparse.mm` can also deal with dense 2nd argument
+        if a.is_sparse and a._nnz() >= a.numel() // 4:
+            a = a.to_dense()
+        a = torch.sparse.mm(matrix, a)
+
+
+def sparse_eye(n: int) -> torch.Tensor:
+    """
+    Create a sparse diagonal matrix.
+
+    .. note ::
+        this is a work-around as long as there is no torch built-in
+
+    :param n:
+        the size
+
+    :return: shape: `(n, n)`, sparse
+        a sparse diagonal matrix
+    """
+    diag_indices = torch.arange(n).unsqueeze(0).repeat(2, 1)
+    return torch.sparse_coo_tensor(indices=diag_indices, values=torch.ones(n))
+
+
+def extract_diagonal_sparse_or_dense(matrix: torch.Tensor, eye: Optional[torch.Tensor]) -> torch.Tensor:
+    """
+    Extract diagonal from a potentially sparse matrix.
+
+    .. note ::
+        this is a work-around as long as `torch.diag` does not work for sparse tensors
+
+    :param matrix: shape: `(n, n)`
+        the matrix
+    :param eye:
+        a prepared sparse eye matrix of appropriate shape
+
+    :return:
+        the diagonal values.
+    """
+    if not matrix.is_sparse:
+        return torch.diag(matrix)
+
+    n = matrix.shape[0]
+    if eye is None:
+        eye = sparse_eye(n=n)
+
+    # we need to use indices here, since there may be zero diagonal entries
+    diag = (matrix * eye).coalesce()
```

Comment:
running the tests on FB15k237, this `coalesce()` call seems to be the most expensive, even more expensive than spmm - I was thinking if this can be improved anyhow, eg, we still can access the COO matrix indices without coalescening, but we can't feed a tensor there, so I tried this
```python
    d = torch.zeros(n)
    for i in range(n):
        d[i] = matrix[i, i]   # torch.sparse.coo only allows direct numbers here, can't feed an eye tensor here
```

Reply:
@mberr could we call for that magic sorcery of torch benchmarking to try out those choices on 5-10 differently sized adj matrices with the sparsity ratio of about fb15k237? 😊

Reply:
The for-loop variant is *much* too slow :sweat_smile:

Reply:
The fastest variant seems to be the "manual coalesce", but this accesses a [private method of sparse tensors](https://pytorch.org/docs/stable/sparse.html#working-with-sparse-coo-tensors) , which itself is considered beta - so this may break easily with torch updates.

Reply:
![times](https://user-images.githubusercontent.com/36984984/167843978-91c76889-5e1b-4255-bbbb-ecbd7c409a8e.svg)

Reply:
worst x axis ever lol, maybe better to log the y axis too?

Reply:
:smile: 

not to mention the heavily overloaded figure, overlap of different line styles, ...

Reply:
benchmarking results are here: https://github.com/pykeen/benchmark-sparse-diagonal

Reply:
wonderful 🤩  manual coalesce on a gpu seems to be of almost constant complexity on variously sized graphs, 500K nodes covers a lot of existing datasets, only WikiKG 2 and Wikidata5M are larger

Reply:
to be fair though - the standard `coalesce` on my M1 CPU did not finish in 30 seconds on FB15k237 :D

Reply:
the manual is quite volatile though - it accesses a private method from an API which is declared to be in beta :smile:

Reply:
some pytorch APIs are in beta for years, so I'd go for this fastest option ;)

Reply:
the test matrices might not be realistic, since they select the non-zero entries randomly. for powers of the adjacency matrix, we might expect to have more non-zero entries on the diagonal (since there are k-circles).

Reply:
i'd also say it's fine to use the "beta" api - same reason as @migalkin